File: games/mafia/backend/app/schemas/ws_events.py
from typing import Literal

from pydantic import BaseModel

from app.schemas.room import RoleOut, RoomOut

# ---- Server -> Client ----
# One class per event `type`. New event types get added here as later steps
# (lobby, gameplay, chat) introduce them — the dispatcher/connection manager
# don't need to change.


class RoomStateEvent(BaseModel):
    """Full room snapshot, sent once right after a socket connects."""

    type: Literal["room_state"] = "room_state"
    room: RoomOut


class PlayerConnectionChangedEvent(BaseModel):
    type: Literal["player_connection_changed"] = "player_connection_changed"
    player_id: str
    connected: bool


class ErrorEvent(BaseModel):
    type: Literal["error"] = "error"
    code: str
    message: str


class PongEvent(BaseModel):
    type: Literal["pong"] = "pong"


class KickedEvent(BaseModel):
    """Sent to a player right before the server closes their socket after
    a host kick, so the client can show why it was disconnected.
    """

    type: Literal["kicked"] = "kicked"


class RoleAssignedEvent(BaseModel):
    """A player's own role, sent only to that player (once at game start,
    and again on reconnect) — never broadcast to the room.
    """

    type: Literal["role_assigned"] = "role_assigned"
    role: RoleOut


# ---- Client -> Server ----
# Parsed by hand in the dispatcher (looking at the raw `type` field) rather
# than a discriminated union — keeps adding a command a one-line diff in the
# dispatcher instead of touching a shared union type. Game/chat commands
# join this list in later steps.


class PingCommand(BaseModel):
    type: Literal["ping"] = "ping"


class SetReadyCommand(BaseModel):
    type: Literal["set_ready"] = "set_ready"
    ready: bool


class UpdateProfileCommand(BaseModel):
    type: Literal["update_profile"] = "update_profile"
    display_name: str | None = None
    avatar: str | None = None


class KickPlayerCommand(BaseModel):
    type: Literal["kick_player"] = "kick_player"
    target_player_id: str


class LeaveRoomCommand(BaseModel):
    type: Literal["leave_room"] = "leave_room"


class StartGameCommand(BaseModel):
    """Host-only: moves the room out of the lobby into the game's first
    phase (Mafia: NIGHT).
    """

    type: Literal["start_game"] = "start_game"


class AdvancePhaseCommand(BaseModel):
    """Host-only: manually advances to the next phase in the game's cycle.
    Stands in for automatic timer/vote-driven advancement (later steps).
    """

    type: Literal["advance_phase"] = "advance_phase"

File: games/spyfall/frontend/types/ws-events.ts
import type { Room } from "./room";

// ---- Server -> Client ----
// Mirrors app/schemas/ws_events.py on the backend, one interface per `type`.

export interface RoomStateEvent {
  type: "room_state";
  room: Room;
}

export interface PlayerConnectionChangedEvent {
  type: "player_connection_changed";
  player_id: string;
  connected: boolean;
}

export interface ErrorEvent {
  type: "error";
  code: string;
  message: string;
}

export interface PongEvent {
  type: "pong";
}

export interface KickedEvent {
  type: "kicked";
}

export interface RoleAssignedEvent {
  type: "role_assigned";
  is_spy: boolean;
  location: string | null;
  role: string | null;
}

export interface VoteCastEvent {
  type: "vote_cast";
  player_id: string;
  target_player_id: string;
}

export interface PlayerRoleRevealOut {
  player_id: string;
  is_spy: boolean;
  role: string | null;
}

export interface GameOverEvent {
  type: "game_over";
  winning_side: string;
  location: string;
  spy_player_ids: string[];
  accused_player_id: string | null;
  reveals: PlayerRoleRevealOut[];
}

export interface DiscussionTimerStartedEvent {
  type: "discussion_timer_started";
  duration_seconds: number;
}

export type ServerEvent =
  | RoomStateEvent
  | PlayerConnectionChangedEvent
  | ErrorEvent
  | PongEvent
  | KickedEvent
  | RoleAssignedEvent
  | VoteCastEvent
  | GameOverEvent
  | DiscussionTimerStartedEvent;

// ---- Client -> Server ----

export interface PingCommand {
  type: "ping";
}

export interface SetReadyCommand {
  type: "set_ready";
  ready: boolean;
}

export interface UpdateProfileCommand {
  type: "update_profile";
  display_name?: string;
  avatar?: string;
}

export interface KickPlayerCommand {
  type: "kick_player";
  target_player_id: string;
}

export interface LeaveRoomCommand {
  type: "leave_room";
}

export interface StartGameCommand {
  type: "start_game";
  enabled_location_keys?: string[] | null;
}

export interface AdvancePhaseCommand {
  type: "advance_phase";
}

export interface CastVoteCommand {
  type: "cast_vote";
  target_player_id: string;
}

export interface GuessLocationCommand {
  type: "guess_location";
  location_key: string;
}

export type ClientCommand =
  | PingCommand
  | SetReadyCommand
  | UpdateProfileCommand
  | KickPlayerCommand
  | LeaveRoomCommand
  | StartGameCommand
  | AdvancePhaseCommand
  | CastVoteCommand
  | GuessLocationCommand;
